taichi/transforms/constant_fold.cpp

```diff
@@ -176,30 +149,76 @@ class ConstantFold : public BasicStmtVisitor {
-        } else if (stmt->cast_type == PrimitiveType::f64) {
-          new_constant.val_f64 = operand->val.val_cast_to_float64();
-        } else {
-          cast_available = false;
-        }
-      }
-      if (cast_available) {
-        insert_and_erase(stmt, new_constant);
-        return;
-      }
-    }
-    auto dst_type = stmt->ret_type;
-    TypedConstant new_constant(dst_type);
-    if (jit_evaluate_unary_op(new_constant, stmt, operand->val)) {
+      new_constant.value_bits = operand->val.value_bits;
       insert_and_erase(stmt, new_constant);
+      return;
     }
+    const auto dt = operand->val.dt;
+    if (!is_good_type(dt))
+      return;
+    const auto dst_type = stmt->ret_type;
+    switch (stmt->op_type) {
+#define HANDLE_REAL_AND_INTEGRAL_UNARY(OP_TYPE, OP_CPP)                     \
+  case UnaryOpType::OP_TYPE: {                                              \
+    if (dt->is_primitive(PrimitiveTypeID::f32) ||                           \
+        dt->is_primitive(PrimitiveTypeID::f64)) {                           \
+      auto res = TypedConstant(dst_type, OP_CPP(operand->val.val_float())); \
+      insert_and_erase(stmt, res);                                          \
+    } else if (dt->is_primitive(PrimitiveTypeID::i32) ||                    \
+               dt->is_primitive(PrimitiveTypeID::i64)) {                    \
+      auto res = TypedConstant(dst_type, OP_CPP(operand->val.val_int()));   \
+      insert_and_erase(stmt, res);                                          \
+    } else if (dt->is_primitive(PrimitiveTypeID::u32) ||                    \
+               dt->is_primitive(PrimitiveTypeID::u64)) {                    \
+      auto res = TypedConstant(dst_type, OP_CPP(operand->val.val_uint()));  \
+      insert_and_erase(stmt, res);                                          \
+    }                                                                       \
+    break;                                                                  \
+  }
+
+      HANDLE_REAL_AND_INTEGRAL_UNARY(neg, -)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(sqrt, std::sqrt)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(round, std::round)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(floor, std::floor)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(ceil, std::ceil)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(abs, std::fabs)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(sin, std::sin)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(asin, std::asin)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(cos, std::cos)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(acos, std::acos)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(tan, std::tan)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(tanh, std::tanh)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(log, std::log)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(exp, std::exp)
+      HANDLE_REAL_AND_INTEGRAL_UNARY(cast_value, )
+      HANDLE_REAL_AND_INTEGRAL_UNARY(rsqrt, 1.0 / std::sqrt)
+#undef HANDLE_REAL_AND_INTEGRAL_UNARY
+
+#define HANDLE_INTEGRAL_UNARY(OP_TYPE, OP_CPP)                             \
```

Comment:
Correct me if were wrong, but looks like `HANDLE_INTEGRAL_UNARY` is a subset of `HANDLE_REAL_AND_INTEGRAL_UNARY`? Why don't we just put `HANDLE_REAL_AND_INTEGRAL_UNARY` here?

Reply:
@jim19930609 the main reason is we actually don't support bit_not/logical_not on real types so I didn't want to fake the support here. (but yea we should have done the type check beforehand...

Reply:
I'm going to send a separate PR for the type check & error message ;)